go/plugins/postgresql/docstore.go

```diff
@@ -118,40 +166,119 @@ func (ds *docStore) Retrieve(ctx context.Context, req *ai.RetrieverRequest) (*ai
 	return res, nil
 }
 
+
+type IndexerOptions struct {
+	VectorStore *VectorStore
+}
+
+
 // Index implements the genkit Retriever.Index method.
 func (ds *docStore) Index(ctx context.Context, req *ai.IndexerRequest) error {
 	if len(req.Documents) == 0 {
 		return nil
 	}
-	//TODO: implement method
-	return nil
+
+
+	/*if req.Options == nil {
+		vs, _ := NewVectorStore()
+
+		req.Options = &IndexerOptions{VectorStore: vs}
+	}
+
+	iopt, ok := req.Options.(*IndexerOptions)
+	if !ok {
+		return fmt.Errorf("postgres.Indexer options have type %T, want %T", req.Options, &IndexerOptions{})
+	}*/
+
+	ereq := &ai.EmbedRequest{
+		Documents: req.Documents,
+		Options:   ds.embedderOptions,
+	}
+	eres, err := ds.embedder.Embed(ctx, ereq)
+	if err != nil {
+		return fmt.Errorf("postgres.Indexer index embedding failed: %v", err)
+	}
+	for _, doc := range req.Documents {
+		// if no metadata provided, initialize with empty map
+		if doc.Metadata == nil {
+			doc.Metadata = make(map[string]any)
+		}
+
+		// generate the id if it's not defined
+		if _, ok := doc.Metadata[ds.idColumn].(string); !ok {
+			doc.Metadata[ds.idColumn] = uuid.New().String()
+		}
+	}
+
+	return ds.index(ctx, eres.Embeddings, req.Documents)
 }
 
-// newDocStore instantiate a docStore
-func newDocStore(ctx context.Context, p *Postgres, cfg Config) (*docStore, error) {
-	p.mu.Lock()
-	defer p.mu.Unlock()
-	if !p.initted {
-		panic("postgres.Init not called")
+func (ds *docStore) index(ctx context.Context, documentEmbeddings []*ai.DocumentEmbedding, documents []*ai.Document) error {
+	b := &pgx.Batch{}
+
+	for i, doc := range documents {
+		embeddingString := vectorToString(documentEmbeddings[i].Embedding)
+		query, values, err := ds.generateAddDocumentsQuery(
+			doc.Metadata[ds.idColumn].(string), doc.Metadata[ds.contentColumn].(string), embeddingString, doc.Metadata)
+		if err != nil {
+			return err
+		}
+		b.Queue(query, values)
 	}
-	if cfg.Name == "" {
-		return nil, fmt.Errorf("name is empty")
+
+	batchResults := ds.engine.Pool.SendBatch(ctx, b)
+	if err := batchResults.Close(); err != nil {
+		return fmt.Errorf("failed to execute batch: %w", err)
 	}
-	if cfg.Embedder == nil {
-		return nil, fmt.Errorf("embedder is required")
+
+	return nil
+
+}
+
+func vectorToString(v []float32) string {
```

Comment:
Please use the PGVector library like LangChainGo https://github.com/averikitsch/langchaingo/blob/ac70ea0e7d0b8ee8b9ba2c9bf2631aa8bbb329f6/vectorstores/cloudsql/vectorstore.go#L99